cou/steps/plan.py

```diff
@@ -491,48 +524,65 @@ def _get_archive_data_steps(analysis_result: Analysis, args: CLIargs) -> list[Pr
     :type args: CLIargs
-    :return: List of post-upgrade steps.
+    :return: List of pre-upgrade steps.
     :rtype: list[PreUpgradeStep]
     """
-    steps = []
-    if args.upgrade_group in {DATA_PLANE, None}:
-        steps.extend(_get_ceph_mon_post_upgrade_steps(analysis_result.apps_control_plane))
-
-    return steps
+    if args.set_noout and args.upgrade_group in {DATA_PLANE, None}:
+        return [
+            PreUpgradeStep(
+                description="Set ceph cluster 'noout' flag before data plane upgrade",
+                coro=ceph.osd_noout(analysis_result.model, enable=True),
+            )
+        ]
+    if not args.set_noout and args.upgrade_group in {DATA_PLANE, None}:
+        PlanStatus.add_message(
+            "Setting noout for ceph cluster during upgrade is optional but recommended. "
+            "You can use `--set-noout` to add this optional step. For more information, "
+            "Please refer to set-noout action: https://charmhub.io/ceph-mon/actions#set-noout",
+            MessageType.WARNING,
+        )
+        return [
+            PreUpgradeStep(
+                description="Verify ceph cluster 'noout' is unset",
+                parallel=False,
+                coro=ceph.assert_osd_noout_state(analysis_result.model, state=False),
+            )
+        ]
+    return []
 
 
-def _get_ceph_mon_post_upgrade_steps(apps: list[OpenStackApplication]) -> list[PostUpgradeStep]:
-    """Get the post-upgrade step for ceph-mon, where we check the require-osd-release option.
+def _get_post_upgrade_steps(analysis_result: Analysis, args: CLIargs) -> list[PostUpgradeStep]:
+    """Get the post-upgrade steps.
 
-    :param apps: List of OpenStackApplication.
-    :type apps: list[OpenStackApplication]
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :param args: CLI arguments
+    :type args: CLIargs
     :return: List of post-upgrade steps.
-    :rtype: list[PreUpgradeStep]
+    :rtype: list[PostUpgradeStep]
     """
-    ceph_mons_apps = [app for app in apps if isinstance(app, CephMon)]
-
-    steps: list[PostUpgradeStep] = []
-    if not ceph_mons_apps:
-        logger.warning("There is no ceph-mon application. Is this a valid OpenStack cloud?")
-        return steps
-
-    for app in ceph_mons_apps:
-        unit = list(app.units.values())[0]  # getting the first unit, since we don't care which one
+    steps = []
+    if args.upgrade_group in {DATA_PLANE, None}:
         steps.append(
             PostUpgradeStep(
-                f"Ensure that the 'require-osd-release' option in '{app.name}' matches the "
-                "'ceph-osd' version",
-                coro=set_require_osd_release_option(unit.name, app.model),
+                "Ensure ceph-mon's 'require-osd-release' option matches the 'ceph-osd' version",
+                coro=ceph.set_require_osd_release_option(analysis_result.model),
             )
         )
-
+        if args.set_noout:
```

Comment:
Now that I'm understanding why it's not showing into the mocked plans this step...

Isn't the process of setting `noout` before upgrading the cloud a recommended way?

If my assumption is right, I think the logic here should be the reverse. By default we set `noout` and in the end we remove. If the user pass the flag than we don't do the recommend way, so this step wouldn't show up.

I think it's a good practice to have the recommended way by default and if you don't want then you pass the flag.

With this implementation if the user doesn't pass the flag it will:

-  Verify ceph cluster 'noout' is unset
-  If is set will raise Application error and stop
- if it's not set will upgrade and not remove the `noout`. Here is the problem to me because we will leave the cloud with `noout`

Reply:
This action may introduce additional risk to the operation, even though it is recommended. After discussing with the operations team, they prefer to have the `noout` option disabled by default.